daemon/setMounts(): do not make /dev/shm ro

kolyshkin opened the pull request:
It has been pointed out that if --read-only flag is given, /dev/shm
also becomes read-only, and it does not make sense, especially for
the case of --ipc private (added in https://github.com/moby/moby/pull/34087).

It is still debatable whether it makes sense to have it read-only
for --ipc shareable|host|container:NNN, but let's follow the --tmpfs
logic here and do NOT make /dev/shm read-only for all the cases.

Fixes #36503

kolyshkin committed:
daemon/setMounts(): do not make /dev/shm ro

It has been pointed out that if --read-only flag is given, /dev/shm
also becomes read-only in case of --ipc private.

This happens because in this case the mount comes from OCI spec
(since commit 7120976d74195), and is a regression caused by that
commit.

The meaning of --read-only flag is to only have a "main" container
filesystem read-only, not the auxiliary stuff (that includes /dev/shm,
other mounts and volumes, --tmpfs, /proc, /dev and so on).

So, let's make sure /dev/shm that comes from OCI spec is not made
read-only.

Fixes: 7120976d74195 ("Implement none, private, and shareable ipc modes")

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

tonistiigi commented:
LGTM

thaJeztah commented:
Discussing in the maintainers meeting; looks ok to make this rw - this should probably be added to the test that was added in https://github.com/moby/moby/pull/28823

kolyshkin commented:
Need to add a test case similar to https://github.com/moby/moby/pull/28823 but for `--ipc private`. The difference in this case is mount comes from the OCI spec.

kolyshkin committed:
daemon/oci_linux_test: add TestIpcPrivateVsReadonly

The test case checks that in case of IpcMode: private and
ReadonlyRootfs: true (as in "docker run --ipc private --read-only")
the resulting /dev/shm mount is NOT made read-only.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
* added a unit test (which is failing before the fixing commit and is passing after)
* modified a commit message to reflect the latest findings about the meaning of `--read-only`

I am a bit concerned about this being a unit test, as I had to mimic the code flow of some actual code instead of calling it; but last time I tried adding an integration test it was strongly suggested that a unit test should be done instead, so I am following the suggestion here.

kolyshkin commented:
z failure (unrelated):

14:21:28 === RUN   TestServiceWithPredefinedNetwork
14:21:41 --- FAIL: TestServiceWithPredefinedNetwork (12.32s)
14:21:41 	daemon.go:283: [ded513e563058] waiting for daemon to start
14:21:41 	daemon.go:315: [ded513e563058] daemon started
14:21:41 	service_test.go:52: timeout hit after 10s: task count at 1 waiting for 0
14:21:41 	daemon.go:273: [ded513e563058] exiting daemon
14:21:41 FAIL

vdemeester reviewed (approved):
LGTM :frog:

thaJeztah commented:
Opened https://github.com/moby/moby/issues/36547 for the flaky `TestServiceWithPredefinedNetwork`
